### PR TITLE
Client server impl

### PR DESCRIPTION
**Pull request by StoyanTinchev**
# Description
Add initial server implementation with cpprestsdk that supports GET/POST/PUT/DELETE queries.

# Test suite
The file ExampleUsage.cpp is test file.

# Environment
- Host machine -> intel core i7-7700HQ 2.80GHz * 8  with  16gb ram
- OS: pop-os
- compiler: GNU gcc-11

--------------------

#### This PR makes changes which ...

- [ ] are thoroughly tested
- [ ] are well documented

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,11 +16,14 @@ add_compile_options(
         -Werror
         -pedantic
         -ggdb3
+		-Wfatal-errors
         -fdiagnostics-color)
 
 enable_testing()
 include(FetchContent)
 
+add_compile_definitions(NDEBUG="true")
```
Comment (boki1) [inline CMakeLists.txt]: Why?

File: CMakeLists.txt [hunk 2]
```diff
@@ -53,7 +60,8 @@ target_link_libraries(LibEugeneCore
         CONAN_PKG::spdlog
         CONAN_PKG::libnop
         CONAN_PKG::andreasbuhr-cppcoro
-        CONAN_PKG::benchmark)
+        CONAN_PKG::benchmark
```
Comment (boki1) [inline CMakeLists.txt:63]: Could you please merge this into a single line diff?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,6 +1,10 @@
 cmake_minimum_required(VERSION 3.19)
 project(eugene)
 
+
+add_subdirectory(src/core/server)
```
Comment (boki1) [inline CMakeLists.txt:5]: Please remove whitespace changes.

File: CMakeLists.txt [hunk 4]
```diff
@@ -12,11 +16,14 @@ add_compile_options(
         -Werror
         -pedantic
         -ggdb3
+		-Wfatal-errors
```
Comment (boki1) [inline CMakeLists.txt]: :+1:

File: src/core/server/handler/handler.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
+#include <handler/handler.h>
+
```
Comment (boki1) [inline src/core/server/handler/handler.cpp]: Please user `PascalCase` for filenames.

File: src/core/server/handler/handler.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+handler::handler(const utility::string_t &url) : m_listener(url) {
+	m_listener.support(
+		methods::GET,
+		[this](auto &&PH1) {
```
Comment (boki1) [inline src/core/server/handler/handler.cpp]: What is `PH1`? Please use a more descriptive name.

File: src/core/server/handler/handler.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	try {
+		task.get();
+	}
+	catch (...) {
+		// TODO: Log the error
+	}
+}
+
+// TODO: remove this
+void display_json(
```
Comment (boki1) [inline src/core/server/handler/handler.cpp]: Maybe you have forgotten to remove this?

File: src/core/server/handler/handler.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	utility::string_t const &prefix) {
+	ucout << prefix << jvalue.serialize() << std::endl;
+}
+
+bool authenticate(const http_headers &headers) {
+	auto auth_header = headers.find(U("authorization"));
+	std::vector<unsigned char> credentials = utility::conversions::from_base64
+		(auth_header->second.substr(6, auth_header->second.size()));
+
+	std::string username{credentials.begin(), std::find(credentials.begin(), credentials.end(), ':')};
```
Comment (boki1) [inline src/core/server/handler/handler.cpp]: The username cannot contain ':'. Is this intentional?
If so, you should probably put a comment which describes that.

File: src/core/server/handler/handler.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	std::string password{std::find(credentials.begin(), credentials.end(), ':') + 1, credentials.end()};
```
Comment (boki1) [inline src/core/server/handler/handler.cpp]: std::find() is used twice where one is sufficient. Please (if std::find(':') is still your desire), cache the result. Find operation is O(n).

File: src/core/server/handler/handler.cpp (hunk 8 above)
Comment (boki1) [inline src/core/server/handler/handler.cpp]: I would seriousliy prefer of this to be std::string_view.

(Not incredibly important but really nice to have.)

File: src/core/server/handler/handler.h [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+#include <cpprest/json.h>
+#include <cpprest/uri.h>
+#include <cpprest/asyncrt_utils.h>
+#include <cpprest/json.h>
+#include <cpprest/filestream.h>
+#include <cpprest/containerstream.h>
+#include <cpprest/producerconsumerstream.h>
+
+#define TRACE(msg)            ucout << msg
+#define TRACE_ACTION(a, k, v) ucout << (a) << " (" << (k) << ", " << (v) << ")\n"
```
Comment (boki1) [inline src/core/server/handler/handler.h]: I do not like these macroses. 

Please consider using `static constexpr` functions.

File: src/core/server/handler/handler.h [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+#define CHECK_ENDPOINT(path, string, authenticate) \
+    if (path == string && authenticate(request.headers()))
+
+using namespace web;
+using namespace http;
+using namespace utility;
+using namespace http::experimental::listener;
+
+class handler {
```
Comment (boki1) [inline src/core/server/handler/handler.h]: `PascalCase`

**Review by boki1 — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (StoyanTinchev) [reply CMakeLists.txt]: But this is already in devel

File: src/core/server/handler/handler.cpp (hunk 7 above)
Comment (StoyanTinchev) [reply src/core/server/handler/handler.cpp]: It will be removed in the next PR

File: src/core/server/handler/handler.cpp (hunk 8 above)
Comment (StoyanTinchev) [reply src/core/server/handler/handler.cpp]: This will be for the next PR

File: src/core/server/handler/handler.cpp (hunk 9 above)
Comment (StoyanTinchev) [reply src/core/server/handler/handler.cpp]: Next PR

File: src/core/server/handler/handler.cpp (hunk 8 above)
Comment (StoyanTinchev) [reply src/core/server/handler/handler.cpp]: It doesn't work with string_view - the constructor isn't the same

File: src/core/server/handler/handler.h (hunk 10 above)
Comment (StoyanTinchev) [reply src/core/server/handler/handler.h]: They are only for test purposes

**Pull request merged**